### ISSUE TITLE
Does nghttp2 lib support request timeout?

### ISSUE THREAD

**Issue by ghost** [nghttp2/nghttp2#666]
I used nghttp2_submit_ping to send a ping frame, but I don't know how to check the ping frame is send OK or not? and how can I cancel a stream correctly.

**Comment by tatsuhiro-t** [comment]
It sounds like title does not match the content, but I'll answer them all.

> Does nghttp2 lib support request timeout?

No.  libnghttp2 itself have no concept of time.  It is a responsibility of application to cancel request if it thinks that it takes very long time.  Usually, you write application using libnghttp2, and you know how long the request have already took.  

> I used nghttp2_submit_ping to send a ping frame, but I don't know how to check the ping frame is send OK or not?

When you call nghttp2_session_mem_send or nghttp2_session_send, libnghttp2 hands out the PING frame serialized in wire format to the application.  It is application's responsibility to actually send these bytes to off to the network.
You can use nghttp2_on_frame_send_callback to know when a PING frame is handed out to the application.

> how can I cancel a stream correctly.

Use nghttp2_submit_rst_stream.

Or are you talking about libnghttp2_asio?

**Comment by benghopy** [comment]
I am talking about nghttp2 low api, I try this: send a ping frame used nghttp2_submit_ping, and if the network is very slowly, the nghttp2_on_frame_send_callback doesn't call for a long time. and I used nghttp2_submit_rst_stream function to cancel the stream, an then call nghttp2_session_set_stream_user_data this function to set the stream_user_data to NULL, but it doesn't work, used nghttp2_session_get_stream_user_data also can get the user_data point.

**Comment by benghopy** [comment]
1. I just used the low api to make http2 request
2. I set the the upload speed to 1 KB/s from the route's configure web, and I found that the callbacks called very slowly. and it will block a very long time.

**Comment by tatsuhiro-t** [comment]
> I set the the upload speed to 1 KB/s from the route's configure web, and I found that the callbacks called very slowly. and it will block a very long time.

If network is slow, then it takes time to send HTTP/2 frames, so invocation of callback is delayed.  It sounds natural.  But I may be missing something.

Could you tell us what you expect in this situation, and what you found instead?  I'm really not sure what the problem is.  It would be better to show us source code to explain this issue.

> does nghttp2_session_set_stream_user_data only can called in nghttp2's callbacks?

It can be called outside of callbacks.

**Comment by benghopy** [comment]
after nghttp2_submit_request I get an stream_id, then i use nghttp2_submit_rst_stream to cancel the request. but it doesn't work right. it can't cancel the request right now.

**Comment by tatsuhiro-t** [comment]
I tried 2 cases:

1.  right after stream ID was obtained from nghttp2_submit_request, I called nghttp2_submit_rst_stream().  It actually canceled request.
2. right after receiving response headers from server, I called nghttp2_submit_rst_stream().  It actually canceled request.

So it works for me.  Could you describe more detail about how you know that it does not work.
Small source code to reproduce this issue would be nice.

**Comment by benghopy** [comment]
Yes, It's my problem, I used it not the right way. TKS for you answer.

**Comment by benghopy** [comment]
![use warning](https://cloud.githubusercontent.com/assets/8003871/18044604/46a24f52-6e01-11e6-86d6-16e8dc124c38.png)
If the stream is not open, I call nghttp2_submit_rst_stream maybe not used, and nghttp2_session_set_stream_user_data will failed.

**Comment by tatsuhiro-t** [comment]
We have special handling in this case.  When nghttp2_submit_rst_stream is called before nghttp2_before_frame_send_callback is called (i.e., stream is not opened yet), it cancels request.  In this case, RST_STREAM is not sent to the peer.
nghttp2_session_set_stream_user_data fails, yes.

**Comment by ranjithpp** [comment]
what is exactly meant by cancel stream? is it expected to call rst_stream or can it be closed locally by not responding to client?
